phygnn/utilities/tf_utilities.py

```diff
@@ -19,3 +20,174 @@ def tf_log10(x):
+
+    Parameters
+    ----------
+    x : tf.Tensor
+        Tensor with NaNs and shape (n_obs, spatial_1, spatial_2) or
+        (n_obs, spatial_1, spatial_2, n_temporal)
+    mask : tf.Tensor
+        Boolean mask of the non-NaN locations in the original x
+
+    Returns
+    -------
+    x : tf.Tensor
+        Input tensor with NaN values filled with the mean of the non-NaN
+        values or 0 if all values are NaN.
+    mask : tf.Tensor
+        Mask of the input tensor where True is not NaN and False is NaN.
+    """
+    not_nan_float = tf.cast(mask, x.dtype)
+
+    hi_res_zeroed = tf.where(mask, x, tf.zeros_like(x))
+
+    count = tf.reduce_sum(not_nan_float)
+    total = tf.reduce_sum(hi_res_zeroed)
+    mean = tf.math.divide_no_nan(total, count)
+
+    return tf.where(mask, x, mean)
+
+
+def mean_fill(x):
+    """Fill NaN values in the input tensor with the mean of the non-NaN.
+    If all values are NaN, fill with 0. This is assumed to be either a 4D
+    or 5D tensor, with the trailing feature channel included.
+
+    Parameters
+    ----------
+    x : tf.Tensor
+        Tensor with NaNs and shape (n_obs, spatial_1, spatial_2, n_features) or
+        (n_obs, spatial_1, spatial_2, n_temporal, n_features)
+
+    Returns
+    -------
+    x_filled : tf.Tensor
+        Input tensor with NaN values filled with the mean of the non-NaN
+        values or 0 if all values are NaN.
+    mask : tf.Tensor
+        Mask of the input tensor where 1 is not NaN and 0 is NaN.
+    """
+    mask = tf.math.logical_not(tf.math.is_nan(x))
+    hr_feat = [_mean_fill(x[..., i], mask[..., i]) for i in range(x.shape[-1])]
+    hr_feat = tf.stack(hr_feat, axis=-1)
+    return hr_feat, tf.cast(mask, x.dtype)
+
+
+def _idw_fill(x, coords, mask):
+    """IDW fill for flattened tensors without depth dimension. Assumes input
+    shape is (N), where N is the number of points (H * W).
+
+    Parameters
+    ----------
+    x : tf.Tensor
+        Tensor with NaNs and shape (N)
+    coords: tf.Tensor
+        Coordinates of the points in the image, shape (N, 2) where N is the
+        number of points (H * W)
+    mask : tf.Tensor
+        Mask of the input tensor where True is not NaN and False is NaN.
+
+    Returns
+    -------
+    x_filled: tf.Tensor
+        Flattened tensor with NaNs filled
+    """
+    nan_mask = tf.math.logical_not(mask)
+    valid_coords = tf.boolean_mask(coords, mask)
+    nan_coords = tf.boolean_mask(coords, nan_mask)
+    valid_vals = tf.boolean_mask(x, mask)
+
+    diffs = tf.expand_dims(nan_coords, 1) - tf.expand_dims(valid_coords, 0)
+    dists = tf.norm(diffs, axis=-1)
+
+    weights = 1 / dists
+    weights /= tf.reduce_sum(weights, axis=1, keepdims=True)
+
+    vals = tf.reduce_sum(weights * tf.expand_dims(valid_vals, 0), axis=1)
+
+    filled = tf.where(mask, x, tf.zeros_like(x))
+    return tf.tensor_scatter_nd_update(filled, tf.where(nan_mask), vals)
+
+
+def idw_fill(x, low_mem=True):
```

Comment:
oh man i dont have the brainpower to give this a real math review right now, the best i can do is make sure you're aware of 

NSRDB implementation of IDW based on distances calculated from KDTree (wouldn't have to be geospatial KDTree here): https://github.com/NREL/nsrdb/blob/main/nsrdb/utilities/interpolation.py#L136

Nice OS software for IDW in tensorflow you could use: https://www.sciencedirect.com/science/article/pii/S2352711022001479
https://github.com/ElsevierSoftwareX/SOFTX-D-22-00084/blob/master/src/tfinterpy/idw.py

Reply:
Yeah there's also the IDW in rex regridder ;). I wanted to do something that would use tensorflow methods so I avoided previous approaches with trees. This OS option seems to use that as well. I included a test that checks this against a brute force numpy approach.

Reply:
Oh the OS approach is actually pretty interesting - it sets up IDW as a tf.model with fixed layers. Most of the logic with my implementation is handling the NaNs though, and this would still be needed.

Reply:
Yeah you're probably right that you can't use trees here. Sounds good just wanted to make sure you're aware of other implementations and yeah great you have tests